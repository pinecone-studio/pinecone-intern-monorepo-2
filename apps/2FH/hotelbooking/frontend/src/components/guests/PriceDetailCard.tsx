'use client';
import React from 'react';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { GetRoomQuery } from '@/generated';

type RoomData = Pick<GetRoomQuery['getRoom'], 'pricePerNight'>;

interface PriceDetailCardProps {
  room: RoomData | undefined;
}

const PriceDetailCard = ({ room }: PriceDetailCardProps) => {
  const roomPrice = room?.pricePerNight !== undefined && room?.pricePerNight !== null ? (isNaN(Number(room.pricePerNight)) ? 150000 : Number(room.pricePerNight)) : 150000;
  const taxAmount = Math.round(roomPrice * 0.08); // 8% tax
  const totalPrice = roomPrice + taxAmount;

  return (
    <Card>
      <CardHeader>
        <CardTitle className="text-lg font-semibold">Price Detail</CardTitle>
      </CardHeader>
      <CardContent>
        <div className="space-y-3">
          <div className="flex justify-between items-start">
            <div>
              <p className="text-gray-900">1 night</p>
              <p className="text-sm text-gray-500">{roomPrice.toLocaleString()}₮ per night</p>
            </div>
            <p className="text-gray-900 font-medium">{roomPrice.toLocaleString()}₮</p>
          </div>
          <div className="flex justify-between">
            <p className="text-gray-900">Taxes</p>
            <p className="text-gray-900">{taxAmount.toLocaleString()}₮</p>
          </div>
          <div className="border-t pt-3">
            <div className="flex justify-between">
              <p className="text-gray-900 font-semibold">Total price</p>
              <p className="text-gray-900 font-bold">{totalPrice.toLocaleString()}₮</p>
            </div>
          </div>
        </div>
      </CardContent>
    </Card>
  );
};

export default PriceDetailCard;
