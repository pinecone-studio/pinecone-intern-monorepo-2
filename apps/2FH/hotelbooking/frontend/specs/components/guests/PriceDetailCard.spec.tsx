import React from 'react';
import { render, screen } from '@testing-library/react';
import PriceDetailCard from '@/components/guests/PriceDetailCard';
describe('PriceDetailCard', () => {
  const mockRoom = {
    pricePerNight: 150000,
  };
  const defaultProps = {
    room: mockRoom,
  };
  it('renders price detail card correctly', () => {
    render(<PriceDetailCard {...defaultProps} />);
    expect(screen.getByText('Price Detail')).toBeInTheDocument();
    expect(screen.getByText('1 night')).toBeInTheDocument();
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
    expect(screen.getByText('150,000₮')).toBeInTheDocument();
  });
  it('calculates tax correctly', () => {
    render(<PriceDetailCard {...defaultProps} />);
    expect(screen.getByText('12,000₮')).toBeInTheDocument();
  });
  it('calculates total price correctly', () => {
    render(<PriceDetailCard {...defaultProps} />);
    expect(screen.getByText('162,000₮')).toBeInTheDocument();
  });
  it('displays tax section', () => {
    render(<PriceDetailCard {...defaultProps} />);
    expect(screen.getByText('Taxes')).toBeInTheDocument();
  });
  it('displays total price section', () => {
    render(<PriceDetailCard {...defaultProps} />);
    expect(screen.getByText('Total price')).toBeInTheDocument();
  });
  it('handles room with no price', () => {
    const roomWithNoPrice = {};
    render(<PriceDetailCard room={roomWithNoPrice} />);
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
    expect(screen.getByText('150,000₮')).toBeInTheDocument();
  });
  it('handles room with null price', () => {
    const roomWithNullPrice = { pricePerNight: null };
    render(<PriceDetailCard room={roomWithNullPrice} />);
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
  });
  it('handles room with undefined price', () => {
    const roomWithUndefinedPrice = { pricePerNight: undefined };
    render(<PriceDetailCard room={roomWithUndefinedPrice} />);
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
  });
  it('handles room with zero price', () => {
    const roomWithZeroPrice = { pricePerNight: 0 };
    render(<PriceDetailCard room={roomWithZeroPrice} />);
    expect(screen.getByText('0₮ per night')).toBeInTheDocument();
    const zeroElements = screen.getAllByText('0₮');
    expect(zeroElements.length).toBeGreaterThanOrEqual(3); // Should have at least 3 zero amounts
  });
  it('handles room with negative price', () => {
    const roomWithNegativePrice = { pricePerNight: -1000 };
    render(<PriceDetailCard room={roomWithNegativePrice} />);
    expect(screen.getByText('-1,000₮ per night')).toBeInTheDocument();
    expect(screen.getByText('-1,000₮')).toBeInTheDocument();
    expect(screen.getByText('-80₮')).toBeInTheDocument(); // Tax
    expect(screen.getByText('-1,080₮')).toBeInTheDocument(); // Total
  });
  it('handles room with decimal price', () => {
    const roomWithDecimalPrice = { pricePerNight: 150000.5 };
    render(<PriceDetailCard room={roomWithDecimalPrice} />);
    expect(screen.getByText('150,000.5₮ per night')).toBeInTheDocument();
    expect(screen.getByText('150,000.5₮')).toBeInTheDocument();
  });
  it('handles room with very large price', () => {
    const roomWithLargePrice = { pricePerNight: 1000000 };
    render(<PriceDetailCard room={roomWithLargePrice} />);
    expect(screen.getByText('1,000,000₮ per night')).toBeInTheDocument();
    expect(screen.getByText('1,000,000₮')).toBeInTheDocument();
    expect(screen.getByText('80,000₮')).toBeInTheDocument(); // Tax
    expect(screen.getByText('1,080,000₮')).toBeInTheDocument(); // Total
  });
  it('formats numbers with proper locale formatting', () => {
    const roomWithLargePrice = { pricePerNight: 1234567 };
    render(<PriceDetailCard room={roomWithLargePrice} />);
    expect(screen.getByText('1,234,567₮ per night')).toBeInTheDocument();
    expect(screen.getByText('1,234,567₮')).toBeInTheDocument();
    expect(screen.getByText('98,765₮')).toBeInTheDocument(); // Tax (rounded)
    expect(screen.getByText('1,333,332₮')).toBeInTheDocument(); // Total
  });
  it('renders with proper card structure', () => {
    render(<PriceDetailCard {...defaultProps} />);
    const card = document.querySelector('.rounded-lg.border');
    expect(card).toBeInTheDocument();
  });
  it('renders with proper spacing', () => {
    render(<PriceDetailCard {...defaultProps} />);
    const spaceYContainer = document.querySelector('.space-y-3');
    expect(spaceYContainer).toBeInTheDocument();
  });
  it('renders with proper border styling', () => {
    render(<PriceDetailCard {...defaultProps} />);
    const borderContainer = document.querySelector('.border-t.pt-3');
    expect(borderContainer).toBeInTheDocument();
  });
  it('displays proper font weights', () => {
    render(<PriceDetailCard {...defaultProps} />);
    const totalPriceElement = screen.getByText('162,000₮');
    expect(totalPriceElement).toHaveClass('font-bold');
    const totalPriceLabel = screen.getByText('Total price');
    expect(totalPriceLabel).toHaveClass('font-semibold');
  });
  it('handles room with string price', () => {
    const roomWithStringPrice = { pricePerNight: '150000' as any };
    render(<PriceDetailCard room={roomWithStringPrice} />);
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
  });

  it('handles room with invalid price', () => {
    const roomWithInvalidPrice = { pricePerNight: 'invalid' as any };
    render(<PriceDetailCard room={roomWithInvalidPrice} />);

    // Should use default price when invalid
    expect(screen.getByText('150,000₮ per night')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<PriceDetailCard {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with zero price', () => {
    const roomWithZeroPrice = { pricePerNight: 0 };
    const { container } = render(<PriceDetailCard room={roomWithZeroPrice} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with large price', () => {
    const roomWithLargePrice = { pricePerNight: 1000000 };
    const { container } = render(<PriceDetailCard room={roomWithLargePrice} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
