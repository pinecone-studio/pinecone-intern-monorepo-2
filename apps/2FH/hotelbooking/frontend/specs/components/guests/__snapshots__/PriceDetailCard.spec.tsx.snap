// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceDetailCard matches snapshot 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="tracking-tight text-lg font-semibold"
    >
      Price Detail
    </h3>
  </div>
  <div
    class="p-6 pt-0"
  >
    <div
      class="space-y-3"
    >
      <div
        class="flex justify-between items-start"
      >
        <div>
          <p
            class="text-gray-900"
          >
            1 night
          </p>
          <p
            class="text-sm text-gray-500"
          >
            150,000
            ₮ per night
          </p>
        </div>
        <p
          class="text-gray-900 font-medium"
        >
          150,000
          ₮
        </p>
      </div>
      <div
        class="flex justify-between"
      >
        <p
          class="text-gray-900"
        >
          Taxes
        </p>
        <p
          class="text-gray-900"
        >
          12,000
          ₮
        </p>
      </div>
      <div
        class="border-t pt-3"
      >
        <div
          class="flex justify-between"
        >
          <p
            class="text-gray-900 font-semibold"
          >
            Total price
          </p>
          <p
            class="text-gray-900 font-bold"
          >
            162,000
            ₮
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PriceDetailCard matches snapshot with large price 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="tracking-tight text-lg font-semibold"
    >
      Price Detail
    </h3>
  </div>
  <div
    class="p-6 pt-0"
  >
    <div
      class="space-y-3"
    >
      <div
        class="flex justify-between items-start"
      >
        <div>
          <p
            class="text-gray-900"
          >
            1 night
          </p>
          <p
            class="text-sm text-gray-500"
          >
            1,000,000
            ₮ per night
          </p>
        </div>
        <p
          class="text-gray-900 font-medium"
        >
          1,000,000
          ₮
        </p>
      </div>
      <div
        class="flex justify-between"
      >
        <p
          class="text-gray-900"
        >
          Taxes
        </p>
        <p
          class="text-gray-900"
        >
          80,000
          ₮
        </p>
      </div>
      <div
        class="border-t pt-3"
      >
        <div
          class="flex justify-between"
        >
          <p
            class="text-gray-900 font-semibold"
          >
            Total price
          </p>
          <p
            class="text-gray-900 font-bold"
          >
            1,080,000
            ₮
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PriceDetailCard matches snapshot with zero price 1`] = `
<div
  class="rounded-lg border bg-card text-card-foreground shadow-sm"
>
  <div
    class="flex flex-col space-y-1.5 p-6"
  >
    <h3
      class="tracking-tight text-lg font-semibold"
    >
      Price Detail
    </h3>
  </div>
  <div
    class="p-6 pt-0"
  >
    <div
      class="space-y-3"
    >
      <div
        class="flex justify-between items-start"
      >
        <div>
          <p
            class="text-gray-900"
          >
            1 night
          </p>
          <p
            class="text-sm text-gray-500"
          >
            0
            ₮ per night
          </p>
        </div>
        <p
          class="text-gray-900 font-medium"
        >
          0
          ₮
        </p>
      </div>
      <div
        class="flex justify-between"
      >
        <p
          class="text-gray-900"
        >
          Taxes
        </p>
        <p
          class="text-gray-900"
        >
          0
          ₮
        </p>
      </div>
      <div
        class="border-t pt-3"
      >
        <div
          class="flex justify-between"
        >
          <p
            class="text-gray-900 font-semibold"
          >
            Total price
          </p>
          <p
            class="text-gray-900 font-bold"
          >
            0
            ₮
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
